snuba/state/cache/redis/backend.py

```diff
@@ -135,9 +135,12 @@ def build_notify_queue_key(task_ident: str) -> str:
         # This updates the stats object and querylog
         record_cache_hit_type(result[0])
 
+        span = Hub.current.scope.span
         if result[0] == RESULT_VALUE:
             # If we got a cache hit, this is easy -- we just return it.
             logger.debug("Immediately returning result from cache hit.")
+            if span:
+                span.set_data("backend", "cache")
```

Comment:
Since you are interested in this instrumentation for a specific use case of the cache (query), I think this should go in the method where we have context on what span was created:
https://github.com/getsentry/snuba/blob/master/snuba/web/db_query.py#L340-L357

You have access to the same data and you know that we are creating a span (if instrumentation is on).
Doing that would allow this caching module (which is very low level and potentially reusable) not to depend on the sentry integration.

Also adding adding arbitrary data to the current span without any context on what the span is about seems a leaky abstraction as you don't know what you are adding this data to. Imagine a call site creates a span with a "backend" key already.  I think we should add data/tags to the spans when we know about the span we are making changes to.

Reply:
I was thinking the same thing but the code in `db_query` doesn't know whether the cache was actually used or not. That's why I put these flags in the cache code instead. I'll try to use a closure to do this but I'm not 100% sure it will work.

Reply:
OK well it seems to work so problem solved.